drivers/dma/dma_sedi.c

```diff
@@ -0,0 +1,582 @@
+/*
+ * Copyright (c) 2018-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT intel_sedi_dma
+
+#include <errno.h>
+#include <stdio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
+#include <string.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/cache.h>
+#include <soc.h>
+
+#include "sedi_driver_dma.h"
+#include "sedi_driver_core.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sedi_dma, CONFIG_DMA_LOG_LEVEL);
+
+#define CACHE_MASK 31
+
+#define GET_MSB(data64) ((uint32_t)(data64 >> 32))
+#define GET_LSB(data64) ((uint32_t)(data64))
+#define BUILD_64_VAR(up32, low32) ((uint64_t)up32 << 32 | (uint64_t)low32)
+
+extern void dma_isr(sedi_dma_t dma_device);
+
+struct dma_sedi_config_info {
+	sedi_dma_t instance; /* Controller instance. */
+	uint8_t chn_num;
+};
+
+struct dma_sedi_driver_data {
+	struct k_sem sema[DMA_CHANNEL_NUM];
+	struct dma_config dma_configs[DMA_CHANNEL_NUM];
+	uint8_t power_status;
+};
+
+#define DEV_NAME(dev) ((dev)->name)
+#define DEV_DATA(dev) ((struct dma_sedi_driver_data *const)(dev)->data)
+#define DEV_CFG(dev) \
+	((const struct dma_sedi_config_info *const)(dev)->config)
+
+/* map dma num to dma driver pointer*/
+static const struct device *dev_map[SEDI_DMA_NUM] = { 0 };
+
+/*
+ * this function will be called when dma transferring is completed
+ * or error happened
+ */
+static void dma_handler(sedi_dma_t dma_device, int channel, int event_id,
+			void *args)
+{
+	ARG_UNUSED(args);
+	const struct device *dev = dev_map[dma_device];
```

Comment:
remove dev_map, get dev by function parameter args instead.

Reply:
This will change a lot, not just in this patch. It also contains other previous check in files.

Reply:
just replace NULL with dev on line 321 "sedi_dma_init(info->instance, (int)channel, cb, NULL);", I think.